File: lingua-006/src/main/scala/Substitution.scala
package codecamp

case class Substitution(solutions: Map[Type.Var, Type]) {
  def apply(constraints: Set[Constraint]): Set[Constraint] = {
    constraints.map(c => apply(c))
  }

  def apply(constraint: Constraint): Constraint = {
    Constraint(
      apply(constraint.a),
      apply(constraint.b)
    )
  }

  def apply(ty: Type): Type = {
    solutions.foldLeft(ty) { (result, solution) =>
      val (tvar, solutionType) = solution
      substitute(result, tvar, solutionType)
    }
  }

  def substitute(ty: Type, tvar: Type.Var, replacement: Type): Type = {
    ty match {
      case Type.INT => ty
      case Type.BOOL => ty
      case Type.FUN(paramTy, returnTy) =>
        Type.FUN(
          substitute(paramTy, tvar, replacement),
          substitute(returnTy, tvar, replacement)
        )
      case Type.VAR(tvar2) =>
        if (tvar == tvar2) replacement else ty
    }
  }

  def compose(other: Substitution): Substitution = {
    val substitutedThis = solutions.mapValues(s => other.apply(s))
    Substitution(substitutedThis ++ other.solutions)
  }
}

object Substitution {
  def empty = Substitution(Map.empty)

  def fromPair(tvar: Type.Var, ty: Type): Substitution = {
    Substitution(Map(tvar -> ty))
  }
}

File: lingua-007/SwiftML/substitution.swift
// -------------------------------------------------------------------------- //
// SwiftML                                                                    //
// -------------------------------------------------------------------------- //

struct Substitution {
  let solutions: [Int : Type]

  static let empty = Substitution(solutions: [:])

  func applyTo(constraints: Set<Constraint>) -> Set<Constraint> {
    return Set(constraints.map(applyTo(constraint:)))
  }

  func applyTo(constraint: Constraint) -> Constraint {
    switch constraint {
    case let .Equal(a, b):
      return .Equal(applyTo(type: a), applyTo(type: b))
    }
  }

  func applyTo(type: Type) -> Type {
    return solutions.reduce(type) { (result, solution) in
      let (tvar, solutionType) = solution
      return result.substitute(tvar: tvar, with: solutionType)
    }
  }

  func applyTo<A>(term: Term<(A, Type)>) -> Term<(A, Type)> {
    return term.mapAttr { (a, type) in (a, applyTo(type: type)) }
  }

  func compose(with other: Substitution) -> Substitution {
    var substituted = solutions.mapValues { other.applyTo(type: $0) }

    for (tvar, type) in other.solutions {
      substituted[tvar] = type
    }

    return Substitution(solutions: substituted)
  }
}
